lib/Dialect/FIRRTL/Transforms/HoistPassthrough.cpp

```diff
@@ -0,0 +1,571 @@
+    dest.replaceAllUsesWith(temp.getDataRaw());
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// RefDriver implementation.
+//===----------------------------------------------------------------------===//
+
+static RefDefineOp getRefDefine(Value result) {
+  for (auto *user : result.getUsers()) {
+    if (auto rd = dyn_cast<RefDefineOp>(user); rd && rd.getDest() == result)
+      return rd;
+  }
+  return {};
+}
+
+RefDriver RefDriver::get(Value v) {
+  auto refVal = dyn_cast<RefValue>(v);
+  if (!refVal)
+    return {};
+
+  auto rd = getRefDefine(v);
+  if (!rd)
+    return {};
+
+  auto ref = getFieldRefFromValue(rd.getSrc(), true);
+  if (!ref)
+    return {};
+
+  return RefDriver(rd, ref);
+}
+
+Value RefDriver::remat(PortMappingFn mapPortFn, ImplicitLocOpBuilder &builder) {
+  auto mappedSource = mapPortFn(getIndex(source.getValue()));
+  auto newVal = getValueByFieldID(builder, mappedSource, source.getFieldID());
+  auto destType = getDest().getType();
+  if (newVal.getType() != destType)
+    newVal = builder.create<RefCastOp>(destType, newVal);
+  return newVal;
+}
+
+//===----------------------------------------------------------------------===//
+// HWDriver implementation.
+//===----------------------------------------------------------------------===//
+
+static bool hasDontTouchOrInnerSymOnResult(Operation *op) {
+  if (AnnotationSet::hasDontTouch(op))
+    return true;
+  auto symOp = dyn_cast<hw::InnerSymbolOpInterface>(op);
+  return symOp && symOp.getTargetResultIndex() && symOp.getInnerSymAttr();
+}
+
+static bool hasDontTouchOrInnerSymOnResult(Value value) {
+  if (auto *op = value.getDefiningOp())
+    return hasDontTouchOrInnerSymOnResult(op);
+  auto arg = dyn_cast<BlockArgument>(value);
+  auto module = cast<FModuleOp>(arg.getOwner()->getParentOp());
+  return (module.getPortSymbolAttr(arg.getArgNumber())) ||
+         AnnotationSet::forPort(module, arg.getArgNumber()).hasDontTouch();
+}
+
+HWDriver HWDriver::get(Value v) {
+  auto baseValue = dyn_cast<FIRRTLBaseValue>(v);
+  if (!baseValue)
+    return {};
+
+  // Output must be passive, for flow reasons.
+  // Reject aggregates for now, to be conservative re:aliasing writes/etc.
+  // before ExpandWhens.
+  if (!baseValue.getType().isPassive() || !baseValue.getType().isGround())
+    return {};
+
+  auto connect = getSingleConnectUserOf(v);
+  if (!connect)
+    return {};
+  auto ref = getFieldRefFromValue(connect.getSrc());
+  if (!ref)
+    return {};
+
+  // Reject if not all same block.
+  if (v.getParentBlock() != ref.getValue().getParentBlock() ||
+      v.getParentBlock() != connect->getBlock())
+    return {};
+
+  // Reject if cannot reason through this.
+  // Use local "hasDontTouch" to distinguish inner symbols on results
+  // vs on the operation itself (like an instance).
+  if (hasDontTouchOrInnerSymOnResult(v) ||
+      hasDontTouchOrInnerSymOnResult(ref.getValue()))
```

Comment:
Note that we WILL hoist signals that pass through storage with an annotation -- see "HWAnno" test that demonstrates this.

This is justified by IMCP behavior re:propagating constant values through storage with annotations (but keeping them and what drives them), but worth calling attention to.